src/Raven.Server/Documents/Sharding/Handlers/ShardedOutgoingReplicationHandler.cs

```diff
@@ -22,77 +24,89 @@ namespace Raven.Server.Documents.Sharding.Handlers
 {
     public class ShardedOutgoingReplicationHandler : AbstractOutgoingReplicationHandler<TransactionContextPool, TransactionOperationContext>
     {
-        private readonly int _shardNumber;
         private readonly ShardedDatabaseContext.ShardedReplicationContext _parent;
-        private readonly ReplicationQueue _replicationQueue;
+        private ManualResetEvent _hasBatch = new(false);
+        private ReplicationBatch _batch;
+        private TaskCompletionSource _tcs;
 
         private readonly byte[] _tempBuffer = new byte[32 * 1024];
-        private OutgoingReplicationStatsAggregator _lastStats;
         private long _lastEtag;
 
+        internal DateTime _lastDocumentSentTime;
+        internal long LastEtag => _lastEtag;
+        public string MissingAttachmentMessage { get; set; }
+        public bool MissingAttachmentsInLastBatch { get; set; }
+        public string LastDatabaseChangeVector { get; set; }
+        public string SourceDatabaseId;
+        public long CurrentEtag { get; set; }
 
-        public ShardedOutgoingReplicationHandler(ShardedDatabaseContext.ShardedReplicationContext parent, ShardReplicationNode node, int shardNumber,
-            TcpConnectionInfo connectionInfo, ReplicationQueue replicationQueue) :
-            base(connectionInfo, parent.Server, parent.DatabaseName, node, parent.Context.DatabaseShutdown, parent.Server.ContextPool)
+        public ShardedOutgoingReplicationHandler(ShardedDatabaseContext.ShardedReplicationContext parent, ShardReplicationNode node, TcpConnectionInfo connectionInfo, string sourceDatabaseId) :
+            base(connectionInfo, parent.Server, parent.Context.DatabaseName, node, parent.Context.DatabaseShutdown, parent.Server.ContextPool)
         {
             _parent = parent;
-            _shardNumber = shardNumber;
-
             _tcpConnectionOptions = new TcpConnectionOptions
             {
-                DatabaseContext = parent.Context,
+                DatabaseContext = _parent.Context,
                 Operation = TcpConnectionHeaderMessage.OperationTypes.Replication
             };
 
-            _replicationQueue = replicationQueue;
+            SourceDatabaseId = sourceDatabaseId;
         }
 
         protected override void Replicate()
         {
-            while (_cts.IsCancellationRequested == false)
+            while (_cts.Token.IsCancellationRequested == false)
             {
-                while (_replicationQueue.Items[_shardNumber].TryTake(out var items))
+                var hasBatch = _hasBatch.WaitOne(3000); //TODO: time for that
+                _hasBatch.Reset();
```

Comment:
this can be racy, you just timeouted, but before the `Reset` you got a batch... so you will reset it

Reply:
Done